File: kv_store/set_value.py
# set_value.py
from kv_store import kv_store
from lock_manager import kv_lock
import threading

# Helper function to handle retries for setting a key-value pair
def handle_set_thread(key, value, timeout=0.01):
    result = [None]
    # Call thread, append results to result
    thread = threading.Thread(target=lambda: result.append(set_value(key, value)))
    thread.start()
    thread.join(timeout)

    # If timeout occurs, send another thread out
    if thread.is_alive():
        thread = threading.Thread(target=lambda: result.append(set_value(key, value)))
        thread.start()
        thread.join(timeout)
    # Both threads failed - return failure
    if thread.is_alive():
        return None
    return result[1]

def set_value(key, value):
    # Safely set key-value pair using lock
    with kv_lock:
        kv_store.set(key, value)
        # TODO: Forward to next in chain
        return 0

File: Rust/src/set_value.rs
// set_value.rs
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;
use crate::kv_store::KVStore;

pub fn handle_set_thread(
    key: String, 
    value: String, 
    kv_store: Arc<Mutex<KVStore>>, 
    timeout: f64
) -> Option<i32> {
    let key_clone = key.clone();
    let value_clone = value.clone();
    let store_clone = Arc::clone(&kv_store);

    // First attempt
    let handle = thread::spawn(move || {
        set_value(key_clone, value_clone, store_clone)
    });

    match handle.join_timeout(Duration::from_secs_f64(timeout)) {
        Ok(result) => Some(result),
        Err(_) => {
            // Second attempt
            let key_clone = key.clone();
            let value_clone = value.clone();
            let store_clone = Arc::clone(&kv_store);
            let handle = thread::spawn(move || {
                set_value(key_clone, value_clone, store_clone)
            });

            match handle.join_timeout(Duration::from_secs_f64(timeout)) {
                Ok(result) => Some(result),
                Err(_) => None,
            }
        }
    }
}

pub fn set_value(key: String, value: String, kv_store: Arc<Mutex<KVStore>>) -> i32 {
    // Safely set key-value pair using lock
    let mut store = kv_store.lock().unwrap();
    store.set(key, value);
    // TODO: Forward to next in chain
    0
}
